File: intel_briefing/sensors/v2ex.py
# ABOUTME: V2EX sensor using the public hot topics JSON API.
# ABOUTME: Fetches trending topics from the V2EX Chinese tech community; no authentication required.
import logging

import httpx

from intel_briefing.models import ConfigSettings, IntelItem

logger = logging.getLogger(__name__)

V2EX_HOT_API = "https://www.v2ex.com/api/topics/hot.json"


class V2EXSensor:
    """Sensor that fetches hot topics from V2EX via the public JSON API."""

    sensor_name: str = "v2ex"

    def fetch(self, config: ConfigSettings, limit: int) -> list[IntelItem]:
        """Fetch hot topics from V2EX.

        Args:
            config: Full application settings (unused for this sensor — no auth required).
            limit: Maximum number of IntelItem objects to return.

        Returns:
            List of IntelItem objects. Returns an empty list on any failure.
        """
        try:
            headers = {"User-Agent": "Intel-Briefing/1.0"}
            resp = httpx.get(
                V2EX_HOT_API,
                headers=headers,
                timeout=15,
                follow_redirects=True,
            )
            resp.raise_for_status()
            data = resp.json()
        except Exception as exc:
            logger.warning("V2EX: request failed: %s", exc)
            return []

        items: list[IntelItem] = []
        try:
            for topic in data[:limit]:
                replies = topic.get("replies", 0)
                topic_id = str(topic.get("id", ""))
                items.append(
                    IntelItem(
                        id=f"v2ex-{topic_id}",
                        source="v2ex",
                        title=topic.get("title", ""),
                        url=topic.get("url", ""),
                        heat=f"{replies} replies",
                    )
                )
        except Exception as exc:
            logger.warning("V2EX: failed to parse response: %s", exc)
            return []

        return items

File: frontend/src/lib/sensors/v2ex.ts
// ABOUTME: V2EX sensor using the public hot topics JSON API.
// ABOUTME: Fetches trending topics from the V2EX Chinese tech community; no authentication required.
import type { ConfigSettings, IntelItem } from '../models'

const V2EX_HOT_API = 'https://www.v2ex.com/api/topics/hot.json'

export async function fetchV2ex(_config: ConfigSettings, limit: number): Promise<IntelItem[]> {
  try {
    const resp = await fetch(V2EX_HOT_API, {
      headers: { 'User-Agent': 'Intel-Briefing/1.0' },
      signal: AbortSignal.timeout(30000),
    })
    if (!resp.ok) throw new Error(`HTTP ${resp.status} from V2EX`)
    const data = await resp.json() as Array<Record<string, unknown>>

    const items: IntelItem[] = []
    for (const topic of data.slice(0, limit)) {
      const replies = topic.replies ?? 0
      const topicId = String(topic.id ?? '')
      items.push({
        id: `v2ex-${topicId}`,
        source: 'v2ex',
        title: String(topic.title ?? ''),
        url: String(topic.url ?? ''),
        heat: `${replies} replies`,
      })
    }
    return items
  } catch (err) {
    throw err instanceof Error ? err : new Error(String(err))
  }
}
